Fix CSV encoding with pandas 1.0

With pandas 1.0.0rc0, the test `dask/dataframe/io/tests/test_csv.py::test_encoding_gh601[utf-16]` is failing

```pytb
pytest dask/dataframe/io/tests/test_csv.py -k test_encoding_gh601[utf-16] -v
================================================================================ test session starts ================================================================================
platform darwin -- Python 3.7.6, pytest-5.2.1, py-1.8.0, pluggy-0.13.0 -- /Users/taugspurger/Envs/pandas-dev/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/taugspurger/sandbox/dask/.hypothesis/examples')
rootdir: /Users/taugspurger/sandbox/dask, inifile: setup.cfg
plugins: xdist-1.29.0, hypothesis-4.36.2, forked-1.0.2, repeat-0.8.0, cov-2.8.1
collected 122 items / 121 deselected / 1 selected

dask/dataframe/io/tests/test_csv.py::test_encoding_gh601[utf-16] FAILED                                                                                                       [100%]
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

encoding = 'utf-16'

    @pytest.mark.parametrize("encoding", ["utf-16", "utf-16-le", "utf-16-be"])
    def test_encoding_gh601(encoding):
        ar = pd.Series(range(0, 100))
        br = ar % 7
        cr = br * 3.3
        dr = br / 1.9836
        test_df = pd.DataFrame({"a": ar, "b": br, "c": cr, "d": dr})

        with tmpfile(".csv") as fn:
            test_df.to_csv(fn, encoding=encoding, index=False)

            a = pd.read_csv(fn, encoding=encoding)
            d = dd.read_csv(fn, encoding=encoding, blocksize=1000)
>           d = d.compute()

dask/dataframe/io/tests/test_csv.py:1055:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
dask/base.py:165: in compute
    (result,) = compute(self, traverse=False, **kwargs)
dask/base.py:436: in compute
    results = schedule(dsk, keys, **kwargs)
dask/threaded.py:81: in get
    **kwargs
dask/local.py:486: in get_async
    raise_exception(exc, tb)
dask/local.py:316: in reraise
    raise exc
dask/local.py:222: in execute_task
    result = _execute_task(task, data)
dask/core.py:118: in _execute_task
    args2 = [_execute_task(a, cache) for a in args]
dask/core.py:118: in <listcomp>
    args2 = [_execute_task(a, cache) for a in args]
dask/core.py:119: in _execute_task
    return func(*args2)
dask/utils.py:29: in apply
    return func(*args, **kwargs)
dask/dataframe/io/csv.py:68: in pandas_read_text
    df = reader(bio, **kwargs)
../pandas/pandas/io/parsers.py:675: in parser_f
    return _read(filepath_or_buffer, kwds)
../pandas/pandas/io/parsers.py:448: in _read
    parser = TextFileReader(fp_or_buf, **kwds)
../pandas/pandas/io/parsers.py:879: in __init__
    self._make_engine(self.engine)
../pandas/pandas/io/parsers.py:1111: in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
../pandas/pandas/io/parsers.py:1883: in __init__
    self._reader = parsers.TextReader(src, **kwds)
pandas/_libs/parsers.pyx:529: in pandas._libs.parsers.TextReader.__cinit__
    ???
pandas/_libs/parsers.pyx:719: in pandas._libs.parsers.TextReader._get_header
    ???
pandas/_libs/parsers.pyx:915: in pandas._libs.parsers.TextReader._tokenize_rows
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   UnicodeDecodeError: 'utf-16-le' codec can't decode byte 0x0a in position 0: truncated data

pandas/_libs/parsers.pyx:2062: UnicodeDecodeError

```

I haven't been able to fix this yet. A few tidbits

* Pandas fixed an issue with CSV IO and encoding in https://github.com/pandas-dev/pandas/issues/24130. Maybe related
* https://github.com/dask/dask/blob/5db6dad7947f186a9764cdf409b8f50bf7c38f22/dask/dataframe/io/csv.py#L397 looks fishy. That should presumably be encoded with the user-provided encoding?

I'm not sure, but I think there's a issue somewhere in `bytes/core.py` or fsspec's bytes handling

With the following debug code,

```diff
diff --git a/dask/dataframe/io/csv.py b/dask/dataframe/io/csv.py
index 6999fd95b..4033e2dce 100644
--- a/dask/dataframe/io/csv.py
+++ b/dask/dataframe/io/csv.py
@@ -404,6 +404,11 @@ def read_pandas(
         include_path=include_path_column,
         **(storage_options or {})
     )
+    for obj in b_out[1][0]:
+        print(obj)
+        data = obj.compute()
+        data.decode(kwargs.get("encoding"))
+    # breakpoint()
 
     if include_path_column:
         b_sample, values, paths = b_out
diff --git a/dask/dataframe/io/tests/test_csv.py b/dask/dataframe/io/tests/test_csv.py
index fbf0bec12..fdef311a7 100644
--- a/dask/dataframe/io/tests/test_csv.py
+++ b/dask/dataframe/io/tests/test_csv.py
@@ -1039,8 +1039,10 @@ def test_read_csv_with_datetime_index_partitions_n():
         assert_eq(df, ddf)
 
 
-@pytest.mark.parametrize("encoding", ["utf-16", "utf-16-le", "utf-16-be"])
+@pytest.mark.parametrize("encoding", ["utf-8", "utf-16", "utf-16-le", "utf-16-be"])
 def test_encoding_gh601(encoding):
+    import dask
+    dask.config.set(scheduler="single-threaded")
     ar = pd.Series(range(0, 100))
     br = ar % 7
     cr = br * 3.3

```

we do fine for utf-8, but fail for utf-16

```
>           data.decode(kwargs.get("encoding"))
E           UnicodeDecodeError: 'utf-16-le' codec can't decode byte 0x0a in position 1046: truncated data
```

I think somewhere in the blocksize splitting, we've split in the middle of a two byte character sequence. Indeed, I think the fact that we have an odd number of bytes

```
(Pdb) pp len(data)
1047
```

indicates that we've messed something up. IIUC, all utf-16 encoded characters will be 2 or 4 bytes, so we should have an even length bytestring.

--- a/dask/dataframe/io/csv.py
+++ b/dask/dataframe/io/csv.py
@@ -484,6 +484,15 @@ def read_pandas(
         kwargs["lineterminator"] = lineterminator
     else:
         lineterminator = "\n"
+    if "encoding" in kwargs:
+        b_lineterminator = lineterminator.encode(kwargs["encoding"])
+        empty_blob = "".encode(kwargs["encoding"])
+        if empty_blob:
+            # This encoding starts with a Byte Order Mark (BOM), so strip that from the
+            # start of the line terminator, since this value is not a full file.
+            b_lineterminator = b_lineterminator[len(empty_blob) :]
+    else:
+        b_lineterminator = lineterminator.encode()
     if include_path_column and isinstance(include_path_column, bool):
         include_path_column = "path"
     if "index" in kwargs or (
@@ -559,7 +568,6 @@ def read_pandas(
             "Setting ``sample=blocksize``"
         )
         sample = blocksize
-    b_lineterminator = lineterminator.encode()
     b_out = read_bytes(
         urlpath,
         delimiter=b_lineterminator,

--- a/dask/dataframe/io/tests/test_csv.py
+++ b/dask/dataframe/io/tests/test_csv.py
@@ -1155,17 +1155,8 @@ def test_read_csv_with_datetime_index_partitions_n():
         assert_eq(df, ddf)
 
 
-xfail_pandas_100 = pytest.mark.xfail(reason="https://github.com/dask/dask/issues/5787")
-
-
-@pytest.mark.parametrize(
-    "encoding",
-    [
-        pytest.param("utf-16", marks=xfail_pandas_100),
-        pytest.param("utf-16-le", marks=xfail_pandas_100),
-        "utf-16-be",
-    ],
-)
+# utf-8-sig and utf-16 both start with a BOM.
+@pytest.mark.parametrize("encoding", ["utf-8-sig", "utf-16", "utf-16-le", "utf-16-be"])
 def test_encoding_gh601(encoding):
     ar = pd.Series(range(0, 100))
     br = ar % 7

